lib/adb.js

```diff
@@ -538,14 +538,38 @@ ADB.prototype.installApp = function installApp(deviceId, apkFile, opts, callback
 	} else {
 		this.devices(function (err, devices) {
 			if (err) return callback(err);
-			if (devices.filter(function (d) { return d.id == deviceId }).length != 1) return callback(new Error(__('device not found')));
+
+			devices = devices.filter(function (d) { return d.id == deviceId });
+			if (devices.length != 1) return callback(new Error(__('device not found')));
+			const deviceInfo = devices[0];
 
 			androidDetect(this.config, function (err, results) {
 				if (err) return callback(err);
 
-				opts.logger && opts.logger.trace(__('Executing: %s', [results.sdk.executables.adb, '-s', deviceId, 'install', '-r', '-d', apkFile].join(' ').cyan));
+				// Fetch the device's API Level.
+				let deviceApiLevel = 1;
+				if (deviceInfo.sdk) {
+					const value = parseInt(deviceInfo.sdk);
+					if (!isNaN(value)) {
+						deviceApiLevel = value;
+					}
+				}
+
+				// Set up the 'adb' arguments array.
+				let arguments = [];
+				arguments.push('-s', deviceId);
+				arguments.push('install');
+				arguments.push('-r');
+				if (deviceApiLevel >= 17) {
+					// Set the '-d' argument to install an older APK over a new one.
+					// Note: Only supported on Android 4.2 (API Level 17) and higher.
+					arguments.push('-d');
```

Comment:
Why are we keeping the `-d` flag anyways? According to the `adb` help screen, it says "allow version code downgrade (debuggable packages only)". Do we care?

Reply:
I assume someone must have been burned by this at some point. Doesn't hurt to keep it. But I agree that this should be a rare event.